openprescribing/matrixstore/row_grouper.py

```diff
@@ -76,10 +77,51 @@ def sum(self, matrix):
         # Fast path for the "each group contains only one row" case
         if self._single_row_groups_selector is not None:
             return matrix[self._single_row_groups_selector]
+        # Initialise an array to contain the output
         rows = len(self._group_selectors)
         columns = matrix.shape[1]
-        grouped = numpy.empty((rows, columns), dtype=matrix.dtype)
+        grouped_output = numpy.empty((rows, columns), dtype=matrix.dtype)
```

Comment:
Your reworking isn't quite equivalent in that it returns a `matrix` if the input type is `matrix` whereas the original would always return an `ndarray`. That's why it had the awkward feature of needing two references to the output object, one of which may be a `matrix` or an `ndarray` and the other of which is always an `ndarray`.

The whole problem arises because numpy has this `matrix` class which is now effectively deprecated because `ndarray` can do everything it does but better, but which they can't remove because too much existing code depends on it. In particular, the only sparse representations provided are sparse *matrices* rather than sparse arrays so we're forced to use them (though there's [an issue](https://github.com/scipy/scipy/issues/8162) tracking this). 

I'd like to make sure that whenever we have a dense representation rather than a sparse one it's an `ndarray`. The MatrixSum operation already does this, and I'd like to make sure RowGrouper does too. Given that almost any time we deal with MatrixStore data it's gone through at least one of these operations then most of our code should only ever have to deal with `ndarray`s, although obviously the internals of these two operations will be an exception.

Another layer of awkwardness here is that "matrix" is a much more natural English term than "ndarray" and so it's leaked heavily into my variable and function naming. This wasn't helped by the fact that I didn't really understand the above 3 paragraphs until after I'd written quite a lot of the initial code.

FWIW, the two classes I was thinking of for wrapping up these details were `LabelledArray` and `LabelledSparseArray`. But I'm still struggling with how that's all going to work.

Reply:
I see. I think the comment might as well be more verbose about what the "old class" means.